R/colocalization_heatmap.R

```diff
@@ -88,31 +223,109 @@ ColocalizationHeatmap <- function(
+  # Cap values to legend range
+  plot_data <- plot_data %>%
+    mutate(across(all_of(value_col), ~ case_when(
+      .x < legend_range[1] ~ legend_range[1],
+      .x > legend_range[2] ~ legend_range[2],
+      TRUE ~ .x
+    )))
+
+  # Transform size col
+  if (!is.null(size_col_transform)) {
+    stopifnot(
+      "size_col_transform must be a function" =
+        is.function(size_col_transform)
+    )
+    plot_data <- plot_data %>%
+      mutate(!!sym(size_col) := size_col_transform(!!sym(size_col)))
+    size_col_label <- paste0(size_col, "_transformed")
+  } else {
+    size_col_label <- size_col
   }
 
-  # Set range for heatmap legend
-  if (is.null(legend_range)) {
-    legend_range <- max(abs(plot_data)) * c(-1, 1)
+  # Cast long table to a wide matrix
+  # This is necessary for pheatmap (tiles) but not for the ggplot based version (dots)
+  # However, we also do it conditionally for dots if clustering is enabled to get the
+  # correct order of rows and columns
+  if (type == "tiles" || (cluster_rows || cluster_cols)) {
+    plot_data_wide <-
+      plot_data %>%
+      {
+        if (!is.null(size_col) && (size_col %in% names(.))) {
+          select(., -all_of(size_col))
+        } else {
+          .
+        }
+      } %>%
+      pivot_wider(names_from = "marker_2", values_from = all_of(value_col), values_fill = 0) %>%
+      column_to_rownames("marker_1") %>%
+      as.matrix()
+
+    # Cluster rows for the dots method
+    if (cluster_rows && (type == "dots")) {
+      rows_clust <- dist(plot_data_wide, method = clustering_distance_rows) %>%
+        hclust(method = clustering_method)
+      plot_data <- plot_data %>%
+        mutate(marker_1 = factor(marker_1, levels = rows_clust$labels))
+    }
+
+    # Cluster columns for the dots method
+    if (cluster_cols && (type == "dots")) {
+      cols_clust <- dist(plot_data_wide %>% t(), method = clustering_distance_cols) %>%
+        hclust(method = clustering_method)
+      plot_data <- plot_data %>%
+        mutate(marker_2 = factor(marker_2, levels = cols_clust$labels))
+    }
   }
 
-  plot_data[plot_data < legend_range[1]] <- legend_range[1]
-  plot_data[plot_data > legend_range[2]] <- legend_range[2]
+  # Return plot data if requested
+  if (return_plot_data) {
+    if (type == "dots") {
+      return(plot_data)
+    } else {
+      return(plot_data_wide)
+    }
+  }
 
   # Plot heatmap
-  plot_data %>%
-    as.matrix() %>%
-    ComplexHeatmap::pheatmap(
-      breaks = seq(legend_range[1], legend_range[2], length.out = 101),
-      color = colorRampPalette(colors)(100),
-      heatmap_legend_param = list(title = legend_title),
-      ...
-    )
+  if (type == "dots") {
+    p <- ggplot(
+      plot_data,
+      aes(!!sym(marker1_col),
+        !!sym(marker2_col),
+        fill = !!sym(value_col),
+        size = !!sym(size_col)
+      )
+    ) +
+      geom_point(shape = 21) +
+      scale_size(range = size_range) +
+      scale_y_discrete(limits = rev) +
```

Comment:
Maybe I missed it, but I can't find `rev` :o

Reply:
It's the base function `rev` :-)